File: largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        # sentinel
        heights.append(0)

        # increasing
        ms = []
        n = len(heights)
        best = 0

        # we compute left smallest rect and right smallest rect
        for i in range(n):
            while ms and heights[i] < heights[ms[-1]]:
                # i is right due to loop constraints
                # lsi is top of stack only if stack exists, otherwise latest element (all indices)
                top_i = ms.pop()
                height = heights[top_i]

                lsi = ms[-1] if ms else -1
                width = i - lsi - 1

                best = max(best, height * width)

            ms.append(i)
        
        heights.pop()
        return best

File: largest-rectangle-in-histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        // sentinel
        heights.push_back(0);

        std::vector<int> msi;
        int max = 0;

        for (int i = 0; i < heights.size(); ++i) {
            while (!msi.empty() && heights[i] < heights[msi.back()]) {
                int height_i = msi.back();
                msi.pop_back();

                int left_i = msi.empty() ? -1 : msi.back();
                int h = heights[height_i];
                int w = i - left_i - 1;

                max = std::max(max, h * w);
            }

            msi.push_back(i);
        }

        heights.pop_back();

        return max;
    }
};